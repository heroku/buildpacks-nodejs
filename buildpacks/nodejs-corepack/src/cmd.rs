use heroku_nodejs_utils::vrs::Version;
use libcnb::Env;
use std::{
    path::Path,
    process::{Command, Stdio},
};

#[derive(thiserror::Error, Debug)]
pub(crate) enum Error {
    #[error("Couldn't start corepack command: {0}")]
    Spawn(std::io::Error),
    #[error("Couldn't finish corepack  command: {0}")]
    Wait(std::io::Error),
    #[error("Corepack command finished with a non-zero exit code: {0}")]
    Exit(std::process::ExitStatus),
    #[error("Corepack output couldn't be parsed: {0}")]
    Parse(String),
}

/// Execute `corepack --version` to determine corepack version
pub(crate) fn corepack_version(env: &Env) -> Result<Version, Error> {
    let output = Command::new("corepack")
        .arg("--version")
        .envs(env)
        .stdout(Stdio::piped())
        .spawn()
        .map_err(Error::Spawn)?
        .wait_with_output()
        .map_err(Error::Wait)?;

    output
        .status
        .success()
        .then_some(())
        .ok_or(Error::Exit(output.status))?;
    let stdout = String::from_utf8_lossy(&output.stdout);
    stdout
        .parse()
        .map_err(|_| Error::Parse(stdout.into_owned()))
}

/// Execute `corepack enable` to setup a corepack shim
pub(crate) fn corepack_enable(
    package_manager: &str,
    shim_path: &Path,
    env: &Env,
) -> Result<(), Error> {
    let shim_path_string = shim_path.to_string_lossy();
    let mut process = Command::new("corepack")
        .args([
            "enable",
            "--install-directory",
            &shim_path_string,
            package_manager,
        ])
        .envs(env)
        .spawn()
        .map_err(Error::Spawn)?;
    let status = process.wait().map_err(Error::Wait)?;
    status.success().then_some(()).ok_or(Error::Exit(status))
}

/// Execute `corepack prepare` to install the correct package manager
pub(crate) fn corepack_prepare(env: &Env) -> Result<(), Error> {
    let mut process = Command::new("corepack")
        .arg("prepare")
        .envs(env)
        .spawn()
        .map_err(Error::Spawn)?;
    let status = process.wait().map_err(Error::Wait)?;
    status.success().then_some(()).ok_or(Error::Exit(status))
}
