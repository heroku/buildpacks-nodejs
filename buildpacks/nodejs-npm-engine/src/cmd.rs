use commons::fun_run::{CmdError, CommandWithName};
use heroku_nodejs_utils::vrs::Version;
use libcnb::Env;
use std::process::Command;

#[derive(Debug)]
pub(crate) enum Error {
    NpmVersionCommand(CmdError),
    ParseNpmVersion(String),
    NodeVersionCommand(CmdError),
    ParseNodeVersion(String),
}

pub(crate) fn node_version(env: &Env) -> Result<Version, Error> {
    exec_version_command("node", env)
        .map_err(Error::NodeVersionCommand)
        .and_then(|output| output.parse().map_err(|_| Error::ParseNpmVersion(output)))
}

pub(crate) fn npm_version(env: &Env) -> Result<Version, Error> {
    exec_version_command("npm", env)
        .map_err(Error::NpmVersionCommand)
        .and_then(|output| output.parse().map_err(|_| Error::ParseNodeVersion(output)))
}

fn exec_version_command(program: &str, env: &Env) -> Result<String, CmdError> {
    Command::new(program)
        .args(["--version"])
        .envs(env)
        .named_output()
        .and_then(|output| output.nonzero_captured())
        .map(|output| output.stdout_lossy())
}
