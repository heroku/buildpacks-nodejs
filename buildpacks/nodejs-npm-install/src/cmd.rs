use heroku_nodejs_utils::vrs::Version;
use libcnb::Env;
use std::ffi::OsStr;
use std::path::Path;
use std::process::{Command, ExitStatus, Output, Stdio};

type Result<T> = std::result::Result<T, Error>;

#[derive(Debug)]
pub(crate) enum Error {
    Exit(ExitStatus),
    Parse(String),
    Spawn(std::io::Error),
    Wait(std::io::Error),
}

pub(crate) fn npm_version(env: &Env) -> Result<Version> {
    npm_with_output(["--version"], env).and_then(|output| {
        let stdout = String::from_utf8_lossy(&output.stdout);
        stdout
            .parse::<Version>()
            .map_err(|_| Error::Parse(stdout.into_owned()))
    })
}

pub(crate) fn npm_set_cache_config(env: &Env, location: &Path) -> Result<()> {
    npm(
        [
            "config",
            "set",
            "cache",
            &location.to_string_lossy(),
            "--global",
        ],
        env,
    )
}

pub(crate) fn npm_set_no_audit(env: &Env) -> Result<()> {
    npm(["config", "set", "audit", "false", "--global"], env)
}

pub(crate) fn npm_install(env: &Env) -> Result<()> {
    npm(["ci", "--production=false"], env)
}

pub(crate) fn npm_run(env: &Env, script: &String) -> Result<()> {
    npm(["run", script], env)
}

fn npm<I, S>(args: I, env: &Env) -> Result<()>
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    Command::new("npm")
        .args(args)
        .envs(env)
        .spawn()
        .map_err(Error::Spawn)
        .and_then(|mut child| child.wait().map_err(Error::Wait))
        .and_then(|status| status.success().then_some(()).ok_or(Error::Exit(status)))
}

fn npm_with_output<I, S>(args: I, env: &Env) -> Result<Output>
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    Command::new("npm")
        .args(args)
        .envs(env)
        .stdout(Stdio::piped())
        .spawn()
        .map_err(Error::Spawn)
        .and_then(|child| child.wait_with_output().map_err(Error::Wait))
        .and_then(|output| {
            let status = output.status;
            status
                .success()
                .then_some(output)
                .ok_or(Error::Exit(status))
        })
}
