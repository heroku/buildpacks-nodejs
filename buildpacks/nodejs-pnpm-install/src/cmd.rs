use libcnb::Env;
use std::{path::Path, process::Command};

#[derive(Debug)]
pub(crate) enum Error {
    Spawn(std::io::Error),
    Wait(std::io::Error),
    Exit(std::process::ExitStatus),
}

/// Execute `pnpm install` to install dependencies for a pnpm project.
pub(crate) fn pnpm_install(pnpm_env: &Env) -> Result<(), Error> {
    let mut process = Command::new("pnpm")
        .args(["install", "--frozen-lockfile"])
        .envs(pnpm_env)
        .spawn()
        .map_err(Error::Spawn)?;

    let status = process.wait().map_err(Error::Wait)?;

    status.success().then_some(()).ok_or(Error::Exit(status))
}

/// Execute `pnpm run` commands like `build`.
pub(crate) fn pnpm_run(pnpm_env: &Env, script: &str) -> Result<(), Error> {
    let status = Command::new("pnpm")
        .arg("run")
        .arg(script)
        .envs(pnpm_env)
        .spawn()
        .map_err(Error::Spawn)?
        .wait()
        .map_err(Error::Wait)?;

    status.success().then_some(()).ok_or(Error::Exit(status))
}

/// Execute `pnpm config set store-dir` to set pnpm's addressable store location.
pub(crate) fn pnpm_set_store_dir(pnpm_env: &Env, location: &Path) -> Result<(), Error> {
    let status = Command::new("pnpm")
        .args(["config", "set", "store-dir", &location.to_string_lossy()])
        .envs(pnpm_env)
        .spawn()
        .map_err(Error::Spawn)?
        .wait()
        .map_err(Error::Wait)?;

    status.success().then_some(()).ok_or(Error::Exit(status))
}

/// Execute `pnpm config set virtual-store-dir` to set pnpm's virtual store location.
pub(crate) fn pnpm_set_virtual_dir(pnpm_env: &Env, location: &Path) -> Result<(), Error> {
    let status = Command::new("pnpm")
        .args([
            "config",
            "set",
            "virtual-store-dir",
            &location.to_string_lossy(),
        ])
        .envs(pnpm_env)
        .spawn()
        .map_err(Error::Spawn)?
        .wait()
        .map_err(Error::Wait)?;

    status.success().then_some(()).ok_or(Error::Exit(status))
}

/// Execute `pnpm store prune` to remove unused dependencies from the
/// content-addressable store.
pub(crate) fn pnpm_store_prune(pnpm_env: &Env) -> Result<(), Error> {
    let status = Command::new("pnpm")
        .args(["store", "prune"])
        .envs(pnpm_env)
        .spawn()
        .map_err(Error::Spawn)?
        .wait()
        .map_err(Error::Wait)?;

    status.success().then_some(()).ok_or(Error::Exit(status))
}
