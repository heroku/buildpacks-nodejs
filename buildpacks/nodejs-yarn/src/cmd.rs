use crate::yarn::Yarn;
use heroku_nodejs_utils::vrs::Version;
use libcnb::Env;
use std::{
    path::{Path, PathBuf},
    process::{Command, Stdio},
};

#[derive(thiserror::Error, Debug)]
pub(crate) enum Error {
    #[error("Couldn't start yarn command: {0}")]
    Spawn(std::io::Error),
    #[error("Couldn't finish yarn  command: {0}")]
    Wait(std::io::Error),
    #[error("Yarn command finished with a non-zero exit code: {0}")]
    Exit(std::process::ExitStatus),
    #[error("Yarn output couldn't be parsed: {0}")]
    Parse(String),
}

pub(crate) fn yarn_version(env: &Env) -> Result<Version, Error> {
    let output = Command::new("yarn")
        .arg("--version")
        .envs(env)
        .stdout(Stdio::piped())
        .spawn()
        .map_err(Error::Spawn)?
        .wait_with_output()
        .map_err(Error::Wait)?;

    output
        .status
        .success()
        .then_some(())
        .ok_or(Error::Exit(output.status))?;
    let stdout = String::from_utf8_lossy(&output.stdout);
    stdout
        .parse()
        .map_err(|_| Error::Parse(stdout.into_owned()))
}

pub(crate) fn yarn_get_cache(yarn_line: &Yarn, env: &Env) -> Result<PathBuf, Error> {
    let mut args = vec!["config", "get"];
    if yarn_line == &Yarn::Yarn1 {
        args.push("cache-folder");
    } else {
        args.push("cacheFolder");
    }
    let output = Command::new("yarn")
        .args(args)
        .envs(env)
        .stdout(Stdio::piped())
        .spawn()
        .map_err(Error::Spawn)?
        .wait_with_output()
        .map_err(Error::Wait)?;

    let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
    Ok(stdout.into())
}

pub(crate) fn yarn_set_cache(yarn_line: &Yarn, cache_path: &Path, env: &Env) -> Result<(), Error> {
    let cache_path_string = cache_path.to_string_lossy();
    let mut args = vec!["config", "set"];
    if yarn_line == &Yarn::Yarn1 {
        args.append(&mut vec!["cache-folder", &cache_path_string]);
    } else {
        args.append(&mut vec!["cacheFolder", &cache_path_string]);
    }

    let mut process = Command::new("yarn")
        .args(args)
        .envs(env)
        .spawn()
        .map_err(Error::Spawn)?;
    let status = process.wait().map_err(Error::Wait)?;
    status.success().then_some(()).ok_or(Error::Exit(status))
}

pub(crate) fn yarn_install(
    yarn_line: &Yarn,
    zero_install: bool,
    yarn_env: &Env,
) -> Result<(), Error> {
    let mut args = vec!["install"];
    if yarn_line == &Yarn::Yarn1 {
        args.push("--frozen-lockfile");
    } else {
        args.push("--immutable");
        args.push("--inline-builds");
        if zero_install {
            args.push("--check-cache");
            args.push("--immutable-cache");
        }
    }

    let mut process = Command::new("yarn")
        .args(args)
        .envs(yarn_env)
        .spawn()
        .map_err(Error::Spawn)?;

    let status = process.wait().map_err(Error::Wait)?;

    status.success().then_some(()).ok_or(Error::Exit(status))
}

pub(crate) fn yarn_run(yarn_env: &Env, script: &str) -> Result<(), Error> {
    let status = Command::new("yarn")
        .arg("run")
        .arg(script)
        .envs(yarn_env)
        .spawn()
        .map_err(Error::Spawn)?
        .wait()
        .map_err(Error::Wait)?;

    status.success().then_some(()).ok_or(Error::Exit(status))
}
