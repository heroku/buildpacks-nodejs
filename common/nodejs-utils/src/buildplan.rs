use libcnb_data::buildpack_plan::BuildpackPlan;

#[derive(Debug, Default, PartialEq)]
pub struct NodeBuildScriptsMetadata {
    pub enabled: Option<bool>,
    pub skip_pruning: Option<bool>,
}

pub const NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME: &str = "node_build_scripts";
const NODE_BUILD_SCRIPTS_METADATA_ENABLED_KEY: &str = "enabled";
const NODE_BUILD_SCRIPTS_METADATA_SKIP_PRUNING_KEY: &str = "skip_pruning";

pub fn read_node_build_scripts_metadata(
    buildpack_plan: &BuildpackPlan,
) -> Result<NodeBuildScriptsMetadata, NodeBuildScriptsMetadataError> {
    buildpack_plan
        .entries
        .iter()
        .filter(|entry| entry.name == NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME)
        .try_fold(
            NodeBuildScriptsMetadata::default(),
            |mut node_build_hooks_metadata, entry| {
                match entry.metadata.get(NODE_BUILD_SCRIPTS_METADATA_ENABLED_KEY) {
                    Some(toml::Value::Boolean(enabled)) => {
                        node_build_hooks_metadata.enabled = Some(*enabled);
                    }
                    Some(value) => {
                        Err(NodeBuildScriptsMetadataError::InvalidEnabledValue(
                            value.clone(),
                        ))?;
                    }
                    None => {}
                }
                match entry
                    .metadata
                    .get(NODE_BUILD_SCRIPTS_METADATA_SKIP_PRUNING_KEY)
                {
                    Some(toml::Value::Boolean(enabled)) => {
                        node_build_hooks_metadata.skip_pruning = Some(*enabled);
                    }
                    Some(value) => {
                        Err(NodeBuildScriptsMetadataError::InvalidSkipPruningValue(
                            value.clone(),
                        ))?;
                    }
                    None => {
                        // As the front-end buildpacks are the only ones that are likely to
                        // using the buildplan metadata, if the build scripts are disabled,
                        // there's a good chance that we need to also skip pruning.
                        if node_build_hooks_metadata.enabled == Some(false) {
                            node_build_hooks_metadata.skip_pruning = Some(true);
                        } else {
                            node_build_hooks_metadata.skip_pruning = None;
                        }
                    }
                }
                Ok(node_build_hooks_metadata)
            },
        )
}

#[derive(Debug)]
pub enum NodeBuildScriptsMetadataError {
    InvalidEnabledValue(toml::Value),
    InvalidSkipPruningValue(toml::Value),
}

#[cfg(test)]
mod test {
    use super::*;
    use libcnb_data::buildpack_plan::{BuildpackPlan, Entry};
    use toml::{toml, Table};

    #[test]
    fn read_node_build_scripts_when_buildpack_plan_contains_no_entries() {
        let buildpack_plan = BuildpackPlan { entries: vec![] };
        assert_eq!(
            read_node_build_scripts_metadata(&buildpack_plan).unwrap(),
            NodeBuildScriptsMetadata::default()
        );
    }

    #[test]
    fn read_node_build_scripts_when_entry_is_present_with_no_metadata() {
        let buildpack_plan = BuildpackPlan {
            entries: vec![Entry {
                name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                metadata: Table::new(),
            }],
        };
        assert_eq!(
            read_node_build_scripts_metadata(&buildpack_plan).unwrap(),
            NodeBuildScriptsMetadata::default()
        );
    }

    #[test]
    fn read_node_build_scripts_when_entry_is_present_and_metadata_is_declared() {
        let buildpack_plan = BuildpackPlan {
            entries: vec![Entry {
                name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                metadata: toml! {
                    enabled = false
                    skip_pruning = false
                },
            }],
        };
        assert_eq!(
            read_node_build_scripts_metadata(&buildpack_plan).unwrap(),
            NodeBuildScriptsMetadata {
                enabled: Some(false),
                skip_pruning: Some(false)
            }
        );
    }

    #[test]
    fn read_node_build_scripts_when_multiple_entries_are_present_and_metadata_is_declared() {
        let buildpack_plan = BuildpackPlan {
            entries: vec![
                Entry {
                    name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                    metadata: toml! {
                        enabled = false
                    },
                },
                Entry {
                    name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                    metadata: toml! {
                        enabled = true
                    },
                },
                Entry {
                    name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                    metadata: Table::new(),
                },
            ],
        };
        assert_eq!(
            read_node_build_scripts_metadata(&buildpack_plan).unwrap(),
            NodeBuildScriptsMetadata {
                enabled: Some(true),
                skip_pruning: None
            }
        );
    }

    #[test]
    fn read_node_build_scripts_when_entry_contains_invalid_metadata_for_enabled() {
        let buildpack_plan = BuildpackPlan {
            entries: vec![Entry {
                name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                metadata: toml! {
                    enabled = 0
                },
            }],
        };
        match read_node_build_scripts_metadata(&buildpack_plan).unwrap_err() {
            NodeBuildScriptsMetadataError::InvalidEnabledValue(_) => {}
            e @ NodeBuildScriptsMetadataError::InvalidSkipPruningValue(_) => {
                panic!("Unexpected error: {e:?}")
            }
        }
    }

    #[test]
    fn read_node_build_scripts_when_entry_contains_invalid_metadata_for_skip_pruning() {
        let buildpack_plan = BuildpackPlan {
            entries: vec![Entry {
                name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                metadata: toml! {
                    enabled = false
                    skip_pruning = 0
                },
            }],
        };
        match read_node_build_scripts_metadata(&buildpack_plan).unwrap_err() {
            NodeBuildScriptsMetadataError::InvalidSkipPruningValue(_) => {}
            e @ NodeBuildScriptsMetadataError::InvalidEnabledValue(_) => {
                panic!("Unexpected error: {e:?}")
            }
        }
    }
}
