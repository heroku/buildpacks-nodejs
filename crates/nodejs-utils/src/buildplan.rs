use libcnb_data::buildpack_plan::BuildpackPlan;

#[derive(Debug, Default, PartialEq)]
pub struct NodeBuildScriptsMetadata {
    pub enabled: Option<bool>,
}

pub const NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME: &str = "node_build_scripts";
const NODE_BUILD_SCRIPTS_METADATA_ENABLED_KEY: &str = "enabled";

pub fn read_node_build_scripts_metadata(
    buildpack_plan: &BuildpackPlan,
) -> Result<NodeBuildScriptsMetadata, NodeBuildScriptsMetadataError> {
    buildpack_plan
        .entries
        .iter()
        .filter(|entry| entry.name == NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME)
        .try_fold(
            NodeBuildScriptsMetadata::default(),
            |mut node_build_hooks_metadata, entry| {
                match entry.metadata.get(NODE_BUILD_SCRIPTS_METADATA_ENABLED_KEY) {
                    Some(toml::Value::Boolean(enabled)) => {
                        node_build_hooks_metadata.enabled = Some(*enabled);
                    }
                    Some(value) => {
                        Err(NodeBuildScriptsMetadataError::InvalidEnabledValue(
                            value.clone(),
                        ))?;
                    }
                    None => {}
                }
                Ok(node_build_hooks_metadata)
            },
        )
}

#[derive(Debug)]
pub enum NodeBuildScriptsMetadataError {
    InvalidEnabledValue(toml::Value),
}

#[cfg(test)]
mod test {
    use super::*;
    use libcnb_data::buildpack_plan::{BuildpackPlan, Entry};
    use toml::{toml, Table};

    #[test]
    fn read_node_build_scripts_when_buildpack_plan_contains_no_entries() {
        let buildpack_plan = BuildpackPlan { entries: vec![] };
        assert_eq!(
            read_node_build_scripts_metadata(&buildpack_plan).unwrap(),
            NodeBuildScriptsMetadata::default()
        );
    }

    #[test]
    fn read_node_build_scripts_when_entry_is_present_with_no_metadata() {
        let buildpack_plan = BuildpackPlan {
            entries: vec![Entry {
                name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                metadata: Table::new(),
            }],
        };
        assert_eq!(
            read_node_build_scripts_metadata(&buildpack_plan).unwrap(),
            NodeBuildScriptsMetadata::default()
        );
    }

    #[test]
    fn read_node_build_scripts_when_entry_is_present_and_metadata_is_declared() {
        let buildpack_plan = BuildpackPlan {
            entries: vec![Entry {
                name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                metadata: toml! {
                    enabled = false
                },
            }],
        };
        assert_eq!(
            read_node_build_scripts_metadata(&buildpack_plan).unwrap(),
            NodeBuildScriptsMetadata {
                enabled: Some(false)
            }
        );
    }

    #[test]
    fn read_node_build_scripts_when_multiple_entries_are_present_and_metadata_is_declared() {
        let buildpack_plan = BuildpackPlan {
            entries: vec![
                Entry {
                    name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                    metadata: toml! {
                        enabled = false
                    },
                },
                Entry {
                    name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                    metadata: toml! {
                        enabled = true
                    },
                },
                Entry {
                    name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                    metadata: Table::new(),
                },
            ],
        };
        assert_eq!(
            read_node_build_scripts_metadata(&buildpack_plan).unwrap(),
            NodeBuildScriptsMetadata {
                enabled: Some(true)
            }
        );
    }

    #[test]
    fn read_node_build_scripts_when_entry_contains_invalid_metadata() {
        let buildpack_plan = BuildpackPlan {
            entries: vec![Entry {
                name: NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME.to_string(),
                metadata: toml! {
                    enabled = 0
                },
            }],
        };
        match read_node_build_scripts_metadata(&buildpack_plan).unwrap_err() {
            NodeBuildScriptsMetadataError::InvalidEnabledValue(_) => {}
        }
    }
}
