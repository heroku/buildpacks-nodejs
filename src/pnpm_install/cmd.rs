use bullet_stream::state::SubBullet;
use bullet_stream::{style, Print};
use fun_run::{CmdError, CommandWithName};
use libcnb::Env;
use std::io::Stderr;
use std::{path::Path, process::Command};

/// Execute `pnpm install` to install dependencies for a pnpm project.
pub(crate) fn pnpm_install(
    pnpm_env: &Env,
    mut log: Print<SubBullet<Stderr>>,
) -> Result<Print<SubBullet<Stderr>>, CmdError> {
    log.stream_cmd(
        Command::new("pnpm")
            .args(["install", "--frozen-lockfile"])
            .envs(pnpm_env),
    )?;
    Ok(log)
}

/// Execute `pnpm run` commands like `build`.
pub(crate) fn pnpm_run(
    pnpm_env: &Env,
    script: &str,
    mut log: Print<SubBullet<Stderr>>,
) -> Result<Print<SubBullet<Stderr>>, CmdError> {
    log.stream_with(
        format!("Running {script} script", script = style::value(script)),
        |stdout, stderr| {
            Command::new("pnpm")
                .arg("run")
                .arg(script)
                .envs(pnpm_env)
                .stream_output(stdout, stderr)
        },
    )?;
    Ok(log)
}

/// Execute `pnpm config set store-dir` to set pnpm's addressable store location.
pub(crate) fn pnpm_set_store_dir(pnpm_env: &Env, location: &Path) -> Result<(), CmdError> {
    Command::new("pnpm")
        .args(["config", "set", "store-dir", &location.to_string_lossy()])
        .envs(pnpm_env)
        .named_output()
        .map(|_| ())
}

/// Execute `pnpm config set virtual-store-dir` to set pnpm's virtual store location.
pub(crate) fn pnpm_set_virtual_dir(pnpm_env: &Env, location: &Path) -> Result<(), CmdError> {
    Command::new("pnpm")
        .args([
            "config",
            "set",
            "virtual-store-dir",
            &location.to_string_lossy(),
        ])
        .envs(pnpm_env)
        .named_output()
        .map(|_| ())
}

/// Execute `pnpm store prune` to remove unused dependencies from the
/// content-addressable store.
pub(crate) fn pnpm_store_prune(
    pnpm_env: &Env,
    mut log: Print<SubBullet<Stderr>>,
) -> Result<Print<SubBullet<Stderr>>, CmdError> {
    log.stream_cmd(Command::new("pnpm").args(["store", "prune"]).envs(pnpm_env))?;
    Ok(log)
}
