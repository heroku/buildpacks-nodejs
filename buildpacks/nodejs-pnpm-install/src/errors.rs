use crate::cmd;
use crate::PnpmInstallBuildpackError;
use heroku_nodejs_utils::buildplan::{
    NodeBuildScriptsMetadataError, NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME,
};
use indoc::formatdoc;
use libherokubuildpack::log::log_error;

pub(crate) fn on_error(err: libcnb::Error<PnpmInstallBuildpackError>) {
    match err {
        libcnb::Error::BuildpackError(bp_err) => on_buildpack_error(bp_err),
        libcnb_err => log_error(
            "heroku/nodejs-pnpm internal buildpack error",
            formatdoc! {"
                An unexpected internal error was reported by the framework used
                by this buildpack.

                If the issue persists, consider opening an issue on the GitHub
                repository. If you are unable to deploy to Heroku as a result
                of this issue, consider opening a ticket for additional support.

                Details: {libcnb_err}
            "},
        ),
    };
}

fn on_buildpack_error(bp_err: PnpmInstallBuildpackError) {
    match bp_err {
        PnpmInstallBuildpackError::BuildScript(err) => {
            let (context, details) = get_cmd_error_context(err);
            log_error(
                "heroku/nodejs-pnpm build script error",
                formatdoc! {"
                    There was an error while attempting to run a build script
                    from this project's package.json. {context}

                    Details: {details}
                "},
            );
        }
        PnpmInstallBuildpackError::PackageJson(err) => log_error(
            "heroku/nodejs-pnpm package.json error",
            formatdoc! {"
                There was an error while attempting to parse this project's
                package.json file. Please make sure it is present and properly
                formatted.

                Details: {err:?}
            "},
        ),
        PnpmInstallBuildpackError::PnpmInstall(err) => {
            let (context, details) = get_cmd_error_context(err);
            log_error(
                "pnpm install error",
                formatdoc! {"
                    There was an error while attempting to install dependencies
                    with pnpm. {context}

                    Details: {details}
                "},
            );
        }
        PnpmInstallBuildpackError::PnpmDir(err) => {
            let (context, details) = get_cmd_error_context(err);
            log_error(
                "directory error",
                formatdoc! {"
                    There was an error while attempting to configure a pnpm
                    store to a buildpack layer directory. {context}

                    Details: {details}
                "},
            );
        }
        PnpmInstallBuildpackError::PnpmStorePrune(err) => {
            let (context, details) = get_cmd_error_context(err);
            log_error(
                "store error",
                formatdoc! {"
                    There was an error while attempting to prune the pnpm
                    content-addressable store. {context}

                    Details: {details}
                "},
            );
        }
        PnpmInstallBuildpackError::VirtualLayer(err) => {
            log_error(
                "virtual store layer error",
                formatdoc! {"
                    There was an error while attempting to create the virtual
                    store layer for pnpm's installed dependencies.

                    Details: {err:?}
                "},
            );
        }
        PnpmInstallBuildpackError::NodeBuildScriptsMetadata(err) => {
            let NodeBuildScriptsMetadataError::InvalidEnabledValue(value) = err;
            let value_type = value.type_str();
            log_error(
                format!("metadata error in {NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME} build plan"),
                formatdoc! {"
                    A participating buildpack has set invalid `[requires.metadata]` for the 
                    build plan named `{NODE_BUILD_SCRIPTS_BUILD_PLAN_NAME}`.
                    
                    Expected metadata format:
                    [requires.metadata]
                    enabled = <bool>
                    
                    But was:
                    [requires.metadata]
                    enabled = <{value_type}> 
                "},
            );
        }
    };
}

fn get_cmd_error_context(err: cmd::Error) -> (&'static str, String) {
    match err {
        cmd::Error::Spawn(io_err) => ("The operating system was unable to start the command.", format!("{io_err}")),
        cmd::Error::Wait(io_err) => ("The operating system was unable to wait for the command to finish. It was no longer running.", format!("{io_err}")),
        cmd::Error::Exit(exit_code) => ("The command exited with a non-zero exit code.", format!("Exit code {exit_code}"))
    }
}
