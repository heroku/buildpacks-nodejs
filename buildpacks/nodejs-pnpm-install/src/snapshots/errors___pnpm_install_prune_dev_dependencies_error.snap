---
source: buildpacks/nodejs-pnpm-install/src/errors.rs
---
- Debug Info:
  - Command failed `pnpm prune --prod --ignore-scripts`
    exit status: 1
    stdout: <empty>
    stderr: <empty>

! Failed to prune pnpm dev dependencies
!
! The Heroku Node.js pnpm Install buildpack uses the command `pnpm prune --prod --ignore-scripts` to remove your dev dependencies from the production environment. This command failed and the buildpack cannot continue. See the log output above for more information.
!
! Suggestions:
! - Ensure that this command runs locally without error (exit status = 0).
!
! Use the debug information above to troubleshoot and retry your build.
